File: mi-core/server/src/operator-runtime/policy-guard.ts
import { OperatorTaskInput, PolicyDecision } from './types';

const ALLOWED_MODES = new Set(['READ_ONLY', 'SAFE_WRITE_TEST_ONLY']);
const BLOCKED_MODES = new Set(['PRODUCTION_WRITE', 'FINANCIAL_ACTION', 'SECURITY_ACTION', 'CREDENTIAL_ACTION']);
const BLOCKED_TARGET_PATTERNS: Array<{ rule: string; pattern: RegExp }> = [
  { rule: 'doordash_production', pattern: /doordash/i },
  { rule: 'toast_production', pattern: /toast/i },
  { rule: 'quickbooks_production', pattern: /quickbooks|\.qbw|intuit/i },
  { rule: 'google_business_profile', pattern: /business\.google|google business profile|gbp/i },
  { rule: 'dreamhost', pattern: /dreamhost/i },
  { rule: 'cloudflare', pattern: /cloudflare/i },
  { rule: 'banking', pattern: /bank|banking|stripe|payment|credit union/i },
  { rule: 'payroll', pattern: /payroll|gusto|adp/i },
];
const SAFE_HOST_PATTERNS = [/^https:\/\/example\.com/i, /^https:\/\/www\.example\.com/i, /^http:\/\/127\.0\.0\.1(?::\d+)?/i, /^http:\/\/localhost(?::\d+)?/i];

export function evaluatePolicy(task: OperatorTaskInput): PolicyDecision {
  const matched_rules: string[] = [];
  const targetUrl = task.target?.url || '';
  const category = task.target?.category || '';
  const combined = `${targetUrl} ${category}`;

  if (BLOCKED_MODES.has(task.mode)) {
    matched_rules.push(`blocked_mode:${task.mode}`);
    return { ok: false, status: 'BLOCKED_BY_POLICY', reason: 'Production systems are not allowed in MVP', matched_rules };
  }

  if (!ALLOWED_MODES.has(task.mode)) {
    matched_rules.push(`unknown_mode:${task.mode}`);
    return { ok: false, status: 'BLOCKED_BY_POLICY', reason: 'Mode is not permitted for MVP', matched_rules };
  }

  for (const blocked of BLOCKED_TARGET_PATTERNS) {
    if (blocked.pattern.test(combined)) {
      matched_rules.push(`blocked_target:${blocked.rule}`);
      return { ok: false, status: 'BLOCKED_BY_POLICY', reason: 'Production systems are not allowed in MVP', matched_rules };
    }
  }

  const safeUrl = SAFE_HOST_PATTERNS.some((pattern) => pattern.test(targetUrl));
  if (!safeUrl) {
    matched_rules.push('safe_target_required');
    return {
      ok: false,
      status: 'BLOCKED_BY_POLICY',
      reason: 'Only public safe test targets are allowed in MVP',
      matched_rules,
    };
  }

  matched_rules.push('mode_allowed', 'target_allowed');
  return { ok: true, status: 'ALLOWED', reason: 'Allowed for safe MVP execution', matched_rules };
}

File: computer-operator-foundation/operator-runtime/policy_guard.py
"""
Policy Guard — blocks unsafe targets and enforces approval rules.
"""
import re
from typing import Optional

BLOCKED_TARGETS = [
    "doordash",
    "toast",
    "quickbooks",
    "gbp",          # Google Business Profile
    "dreamhost",
    "cloudflare",
    "banking",
    "payroll",
    "chase",
    "wellsfargo",
    "wells fargo",
    "boa",
    "bank of america",
    "citi",
    "citi.com",
    "intuit.com",
    "turbotax",
    "mint.com",
    "business.google.com",
]

BLOCKED_PATTERNS = [
    r"door\s*dash",
    r"toast\s*(tab|pos|crm)",
    r"quick\s*books",
    r"google\s*business\s*profile",
    r"dream\s*host",
    r"cloud\s*flare",
    r".*bank(ing)?.*",
    r"pay\s*roll",
    r"paychex",
    r"adp\.com",
    r"google\.com/business",
    r"business\.google\.com",
]

APPROVAL_LEVELS = {
    "READ_ONLY",
    "SAFE_WRITE",
    "PRODUCTION_WRITE",
    "FINANCIAL_ACTION",
    "SECURITY_ACTION",
}


def classify_target(target: str) -> str:
    """Classify a target URL/string as SAFE or BLOCKED."""
    t = target.lower().strip()
    for blocked in BLOCKED_TARGETS:
        if blocked in t:
            return "BLOCKED"
    for pattern in BLOCKED_PATTERNS:
        if re.search(pattern, t):
            return "BLOCKED"
    return "SAFE"


def check_target(target: str) -> dict:
    """Check if a target is allowed. Returns policy decision dict."""
    classification = classify_target(target)
    decision = {
        "ok": classification == "SAFE",
        "target": target,
        "classification": classification,
        "status": "APPROVED" if classification == "SAFE" else "BLOCKED_BY_POLICY",
    }
    return decision


def check_approval_level(level: str) -> dict:
    """Validate an approval level."""
    return {
        "valid": level in APPROVAL_LEVELS,
        "level": level,
    }


def is_production_target(target: str) -> bool:
    """Check if a target looks like a production system."""
    t = target.lower()
    production_indicators = ["prod", "production", "live", "real", ".com", ".io", ".net"]
    safe_indicators = ["localhost", "127.0.0.1", "file://", "test", "demo", "example"]
    for s in safe_indicators:
        if s in t:
            return False
    for p in production_indicators:
        if p in t:
            return True
    return False
